File: src/cpp/implementation/inner_set_1.cpp
#include "inner_set_0.hpp"
#include "inner_set_1.hpp"

template<typename T>
inner_set_1<T>::inner_set_1(T p_element)
{
  a_values.element_1 = p_element;
}

template<typename T>
inner_set_1<T>::~inner_set_1()
{

}

template<typename T>
inner_set<T>* inner_set_1<T>::add_element(T p_element)
{
    if(!contains_element(p_element))
        return new inner_set_2<T>(this, p_element);
    else
        return this;
}

template<typename T>
inner_set<T>* inner_set_1<T>::add_all_elements(inner_set<T>* p_inner_set)
{
    return p_inner_set->add_element(a_values.element_1);
}

template<typename T>
bool inner_set_1<T>::contains_element(T p_element)
{
    return  a_values.element_1 == p_element;
}

template<typename T>
bool inner_set_1<T>::contains_all_elements(inner_set<T>* p_inner_set)
{
  return p_inner_set->contains_element(a_values.element_1);
}

template<typename T>
inner_set<T>*  inner_set_1<T>::copy()
{
	return this;
}

template<typename T>
T inner_set_1<T>::get_element(int p_index)
{
    return reinterpret_cast<T*>(&a_values)[p_index];
}

template<typename T>
inner_set<T>* inner_set_1<T>::remove_element(T p_element)
{
  if(a_values.element_1 == p_element)
    return inner_set_0<T>::EMPTY;
  else
    return this;
}

template<typename T>
inner_set<T>* inner_set_1<T>::remove_all_elements(inner_set<T>* p_inner_set)
{
  return p_inner_set->remove_element(a_values.element_1);
}

template<typename T>
inner_set<T>* inner_set_1<T>::retain_all_elements(inner_set<T>* p_inner_set)
{
  inner_set<T>* r_inner_set = inner_set_0<T>::EMPTY;

  if(p_inner_set->contains_element(a_values.element_1))
    r_inner_set = r_inner_set->add_element(a_values.element_1);

  return r_inner_set;
}

template<typename T>
inner_set<T>* inner_set_1<T>::add_and_propagate(T p_element, micro_set<T>* p_micro_set)
{
  if(a_values.element_1 == p_element)
    return this;
  else
  {
	   p_micro_set->insert(p_element);

	    return new inner_set_2<T>(this, p_element);
  }
}

template<typename T>
inner_set<T>* inner_set_1<T>::add_all_and_propagate(inner_set<T>* p_inner_set, micro_set<T>* p_micro_set)
{
  return p_inner_set->add_all_and_propagate_reverse(this, p_micro_set);
}

template<typename T>
inner_set<T>* inner_set_1<T>::add_all_and_propagate_reverse(inner_set<T>* p_inner_set, micro_set<T>* p_micro_set)
{
  return p_inner_set->add_and_propagate(a_values.element_1, p_micro_set);
}

template<typename T>
int inner_set_1<T>::get_size()
{
    return 1;
}

File: src/scala/implementation/InnerSet_1.scala
package scala.implementation


class InnerSet_1[T](protected val element1: T) extends AbstractInnerSet[T]
{

    def this(p_innerSet: InnerSet_1[T]) = this(p_innerSet.element1 )
    override def addElement(p_element: T): InnerSet[T] =
    {
        if (element1.equals(p_element))
        {
            this
        }
        else  new InnerSet_2(element1, p_element)
    }

    override def addUnchecked(p_element: T): InnerSet[T] =  new InnerSet_2(element1, p_element)

    override def addAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =  p_innerSet.addElement(element1)

    override def containsElement(p_element: Object): Boolean = element1.equals(p_element)

    override def containsAllElements(p_innerSet: InnerSet[T]): Boolean =  p_innerSet.containsElement(element1.asInstanceOf[Object])

    override def getElement(p_index: Int): Option[T] = p_index match{
        case 1 => Some(element1)
        case _ => None
    }


    override def removeElement(p_element: Object): InnerSet[T] =
    {
            if(element1.equals(p_element))
         new InnerSet_0
                        else this
    }

    override def removeAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =  p_innerSet.removeElement(element1.asInstanceOf[Object])

    override def retainAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =
    {
      var i: InnerSet[T] = this
            if(!p_innerSet.containsElement(element1.asInstanceOf[Object])) i = i.removeElement(element1.asInstanceOf[Object])
            i
    }

    override def getSize: Int = 1

    override def iterator: InnerSetIterator[T] = new InnerSetIterator(this)

    override def copy: Option[InnerSet[T]] = Some(this)

    override def clear(unused: Boolean): InnerSet[T] = new InnerSet_0[T]

    override def toString : String = "{ " + element1 + " }"

    override def addAndPropagate(p_element: T, p_microSet: MicroSet[T]): InnerSet[T] =
    {
        if (element1.equals(p_element))
        {
            this
        }
        else         {
            p_microSet.add(p_element)
            new InnerSet_2(element1, p_element)
        }    }

    override def addAllAndPropagate(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] = p_innerSet.addAllAndPropagateReverse(this, p_microSet)

    override def addAllAndPropagateReverse(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] =  p_innerSet.addAndPropagate(element1,p_microSet)
}